File: problems/005-transactional-vfs-quota/solutions/cpp/main.cpp
#include <algorithm>
#include <cstdint>
#include <iostream>
#include <string>
#include <vector>
using namespace std;
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int p, n, limit;
  if (!(cin >> p >> n))
    return 0;
  uint64_t cap;
  cin >> cap >> limit;
  vector<char> ex(p + 1);
  vector<uint64_t> sz(p + 1);
  uint64_t used = 0, peakb = 0;
  int ino = 0, peaki = 0, sticky = 0;
  while (n--) {
    string op, err;
    int x;
    cin >> op >> x;
    if (op == "CREATE") {
      if (ex[x])
        err = "EXISTS";
      else if (ino == limit)
        err = "INODES";
      else
        ex[x] = 1, sz[x] = 0, ino++;
    } else if (op == "UNLINK") {
      if (!ex[x])
        err = "NOENT";
      else
        used -= sz[x], sz[x] = 0, ex[x] = 0, ino--;
    } else {
      uint64_t v;
      if (op == "WRITE") {
        uint64_t off, len;
        cin >> off >> len;
        v = len ? max(sz[x], off + len) : sz[x];
      } else
        cin >> v;
      if (!ex[x])
        err = "NOENT";
      else if (v > sz[x] && v - sz[x] > cap - used)
        err = "BYTES";
      else {
        if (v >= sz[x])
          used += v - sz[x];
        else
          used -= sz[x] - v;
        sz[x] = v;
      }
    }
    if (err.empty())
      cout << "OK\n";
    else {
      cout << "ERR " << err << '\n';
      if (err == "BYTES" || err == "INODES")
        sticky = 1;
    }
    peakb = max(peakb, used);
    peaki = max(peaki, ino);
  }
  cout << "SUMMARY " << used << ' ' << ino << ' ' << peakb << ' ' << peaki
       << ' ' << sticky << '\n';
}

File: problems/005-transactional-vfs-quota/solutions/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()
	var p, n, limit int
	var cap uint64
	if _, e := fmt.Fscan(in, &p, &n, &cap, &limit); e != nil {
		return
	}
	ex := make([]bool, p+1)
	sz := make([]uint64, p+1)
	var used, peakb uint64
	ino, peaki, sticky := 0, 0, 0
	for z := 0; z < n; z++ {
		var op string
		var x int
		fmt.Fscan(in, &op, &x)
		err := ""
		if op == "CREATE" {
			if ex[x] {
				err = "EXISTS"
			} else if ino == limit {
				err = "INODES"
			} else {
				ex[x] = true
				ino++
			}
		} else if op == "UNLINK" {
			if !ex[x] {
				err = "NOENT"
			} else {
				used -= sz[x]
				sz[x] = 0
				ex[x] = false
				ino--
			}
		} else {
			var v uint64
			if op == "WRITE" {
				var off, length uint64
				fmt.Fscan(in, &off, &length)
				v = sz[x]
				if length > 0 && off+length > v {
					v = off + length
				}
			} else {
				fmt.Fscan(in, &v)
			}
			if !ex[x] {
				err = "NOENT"
			} else if v > sz[x] && v-sz[x] > cap-used {
				err = "BYTES"
			} else {
				if v >= sz[x] {
					used += v - sz[x]
				} else {
					used -= sz[x] - v
				}
				sz[x] = v
			}
		}
		if err == "" {
			fmt.Fprintln(out, "OK")
		} else {
			fmt.Fprintln(out, "ERR", err)
			if err == "BYTES" || err == "INODES" {
				sticky = 1
			}
		}
		if used > peakb {
			peakb = used
		}
		if ino > peaki {
			peaki = ino
		}
	}
	fmt.Fprintln(out, "SUMMARY", used, ino, peakb, peaki, sticky)
}
